src/Microsoft.AspNet.Authorization/Infrastructure/AssertionRequirement.cs

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNet.Authorization.Infrastructure
+{
+    public class AssertionRequirement : AuthorizationHandler<AssertionRequirement>, IAuthorizationRequirement
+    {
+        /// <summary>
+        /// Function that is called to handle this requirement
+        /// </summary>
+        public Func<AuthorizationContext, bool> Handler { get; }
+
+        public AssertionRequirement(Func<AuthorizationContext, bool> handleMe)
+        {
+            Handler = handleMe;
+        }
+
+        protected override void Handle(AuthorizationContext context, AssertionRequirement requirement)
```

Comment:
We don't have any sugar for adding an async assert today.  I'm not opposed to adding support for a `Func<AuthorizationContext, Task<bool>>` but it will be a bit messier with two of these...  @blowdart thoughts?